src/serializationStrategies/XpathRangeSelector/xpath.spec.js

```diff
@@ -0,0 +1,417 @@
+    const target = document.getElementById('target').previousSibling;
+    const [result] = xpath.getFirstByXPath("./*[name()='span'][1]/*[name()='div'][1]/*[name()='section'][1]/text()[1]", 0, reference);
+
+    expect(result).toEqual(target);
+    expect(result.textContent).toEqual('asdfasdf');
+  });
+
+  it('doesn\'t count highlights between text', () => {
+    document.body.innerHTML = `
+      <div id="reference">
+        <div></div>
+        <div></div>
+        <div>
+          werwerwerwer
+          <div>
+            qwer <span ${DATA_ATTR}>asdf</span> werewwer
+            <div></div>
+            asdfasdf
+          </div>
+        </div>
+        <div></div>
+      </div>
+    `;
+
+    const reference = document.getElementById('reference');
+    const [result] = xpath.getFirstByXPath("./*[name()='div'][3]/*[name()='div'][1]/text()[2]", 0, reference);
+
+    expect(result.textContent.trim()).toEqual('asdfasdf');
+  });
+
+  it('rolls over into subsequent text nodes if offset is too big when they are separated by highlights', () => {
+    document.body.innerHTML = `
+      <div id="reference">
+        <div></div>
+        <div></div>
+        <div>
+          werwerwerwer
+          <div>
+            qwer <span ${DATA_ATTR}>asdf</span> werewwer
+            <div></div>
+            asdfasdf
+          </div>
+        </div>
+        <div></div>
+      </div>
+    `;
+
+    const reference = document.getElementById('reference');
+    const [result, offset] = xpath.getFirstByXPath("./*[name()='div'][3]/*[name()='div'][1]/text()[1]", 23, reference);
+
+    expect(result.textContent.trim()).toEqual('werewwer');
+    expect(offset).toEqual(1);
+  });
+
+  it('a nonexistant offset should return null', () => {
+    document.body.innerHTML = `
+      <div id="reference">asdf<span>asdf</span>qwerqwerqwerqwer</div>
+    `;
+
+    const reference = document.getElementById('reference');
+    const [result] = xpath.getFirstByXPath("./text()[1]", 10, reference);
+
+    expect(result).toEqual(null);
+  });
+
+  it('if the offset specifies the end boundary of highlight and text, it returns the text', () => {
+    document.body.innerHTML = `
+      <div id="reference">
+        qwer <span ${DATA_ATTR}>asdf</span> werewwer
+      </div>
+    `;
+
+    const reference = document.getElementById('reference');
+    const [result] = xpath.getFirstByXPath("./text()[1]", 18, reference);
+
+    expect(result.textContent.trim()).toEqual('werewwer');
+  });
+
+
+  it('if the offset specifies the start boundary of highlight and text, it returns the text', () => {
+    document.body.innerHTML = `
+      <div id="reference">
+        qwer <span ${DATA_ATTR}>asdf</span> werewwer
+      </div>
+    `;
+
+    const reference = document.getElementById('reference');
+    const [result] = xpath.getFirstByXPath("./text()[1]", 14, reference);
+
+    expect(result.textContent.trim()).toEqual('qwer');
+  });
+});
```

Comment:
❤️ SPECS!  Look pretty comprehensive, but reading them I got to wondering: "How do we handle overlapping highlights?"  Is that somewhere in the library, or is the consumer responsible for making sure they don't overlap?

Reply:
yea, the way the library is written overlapping highlights would just not be displayed, we could add some validation that threw an error in the `highlighter.highlight` method